### ISSUE TITLE
[expo-av] Video not reliable loading

### ISSUE THREAD

**Issue by RRaideRR** [expo/expo#20483]
### Summary

A Video is not always loaded. To be more precise, in some cases `onLoadStart` is called, however `onLoad`, `onReadyForDisplay`, `onError` are not called. 

This occurs on iOS and standalone apps. It doesn't matter if its on my local file system or if its a remote video. 

### What platform(s) does this occur on?

iOS

- Tested with iOS 15.7.1 --> not working
- Tested with iOS 15.7.2 --> not working

Does not occur on Android

### Environment

expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.1
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.18.1 - /opt/homebrew/opt/node@16/bin/node
      Yarn: 1.22.19 - ~/.yarn/bin/yarn
      npm: 8.19.2 - /opt/homebrew/opt/node@16/bin/npm
      Watchman: 2022.11.14.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.1, iOS 16.1, macOS 13.0, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: 2021.3 AI-213.7172.25.2113.9123335
      Xcode: 14.1/14B47b - /usr/bin/xcodebuild
    npmPackages:
      babel-preset-expo: ~9.2.1 => 9.2.2 
      expo: ~47.0.8 => 47.0.8 
      react: 18.1.0 => 18.1.0 
      react-dom: 18.1.0 => 18.1.0 
      react-native: 0.70.5 => 0.70.5 
      react-native-web: ~0.18.7 => 0.18.9 
    npmGlobalPackages:
      eas-cli: 3.0.0
      expo-cli: 6.0.8
    Expo Workflow: bare

### Minimal reproducible example

The expo snack has only limited informative value because it seems that this only appears on standalone apps. However, I also attached a video that is showing the faulty behavior.  It doesn't matter if its bare or managed workflow.

Snack: https://snack.expo.dev/@raider/video-issue

Screenrecording from an already cached video: 
https://user-images.githubusercontent.com/3411292/207832104-5f19f9bb-be61-4b9e-a142-c7611ea71003.mp4

Screenrecording from a remote video:
https://user-images.githubusercontent.com/3411292/207834857-9a210792-4144-43fe-98c6-d0420ba629d2.MOV

**Comment by RRaideRR** [comment]
My solution to this problem: 

1. Download it to the cache
2. Load it from the cache
3. Set a timer - if the video does not load within a timeframe (e.g. 500ms), unmount the video and mount it again.

**Comment by RRaideRR** [comment]
This is a simple code snippet that shows how I could solve it:

```
import React, { memo, useCallback, useEffect, useRef } from 'react';

import { Video } from 'expo-av';
import { VideoProps } from 'expo-av/src/Video.types';

/**
 * On iOS, the video does not always reliable load.
 * So we are setting a timer when the load starts, and then if the video is not ready after 300 ms, we are reloading it.
 * https://github.com/expo/expo/issues/20483
 */
function ReloadableVideo(props: VideoProps) {
  const [shouldVideoReload, setShouldVideoReload] = React.useState(false);
  const isVideoReadyRef = useRef<boolean>(false);

  const potentiallyReloadVideo = useCallback(() => {
    setTimeout(() => {
      if (!isVideoReadyRef.current) {
        setShouldVideoReload(true);
      }
    }, 300);
  }, []);

  useEffect(() => {
    if (shouldVideoReload) {
      setShouldVideoReload(false);
    }
  }, [shouldVideoReload]);

  return (
    <>
      {!shouldVideoReload && (
        <Video
          {...props}
          onLoadStart={() => {
            potentiallyReloadVideo();
            if (props.onLoadStart) {
              props.onLoadStart();
            }
          }}
          onReadyForDisplay={(event) => {
            isVideoReadyRef.current = true;
            if (props.onReadyForDisplay) {
              props.onReadyForDisplay(event);
            }
          }}
        />
      )}
    </>
  );
}

export default memo(ReloadableVideo);

```

**Comment by RRaideRR** [comment]
Since this is still an issue and I just provided a (not optimal) workaround, I'd vote for not closing this.

**Comment by mikealche** [comment]
This happens to me as well. Only in production. 
Trying to load videos inside a FlatList (think TikTok style), and some may not load randomly...

**Comment by coofzilla** [comment]
@mikealche This is our exact use case. Video players in a FlatList, and _sometimes_ they'll just stop loading randomly. I checked the playback status, and it doesn't even seem to be triggering. Gonna give this unmounting thing a try and hopefully that works 🤞🏼.

**Comment by RRaideRR** [comment]
Yeah I also investigated in this and couldn't reproduce this error reliable. However, this does not only occur on videos inside a FlatList but also with normal videos (see the minimal reproducible example).

**Comment by quyentho** [comment]
Do you have any updates so far?

**Comment by miso-belica** [comment]
It seems this is the same issue as https://github.com/expo/expo/issues/1131 and related to https://github.com/expo/expo/pull/9443. Unfortunately, expo team has no time to look into it. If someone finds it here maybe an advice is to use some more mature library for video in RN because if you use `expo-av` you can **expect videos not loaded randomly** for your iOS customers 🙁 Sad thing is even tools like Sentry can't inform you about this because `onError` handler is not called when it happens. Only your customers can report it to you.

Thanks @RRaideRR for your hack. I gonna try it too. Maybe I try to look into expo code if I don't notice some shorter workaround but I have never seen iOS code before so maybe a different library is the easiest solution in the long term here.

**Comment by RRaideRR** [comment]
yeah to be honest: its not a perfect solution - more like a dirty hack... if its not going to be rendered in one second, it will try to endless reload the whole thing :-/. Not really happy with the overall solution here. 

@brentvatne probably this has not the highest priority for the expo team right now - right 😬? I can't imagine that someone can create an app that is heavily dependent on videos without being sure that a video has loaded or not.

**Comment by adriaanbalt** [comment]
> @RRaideRR  probably this has not the highest priority for the expo team right now - right 😬? I can't imagine that someone can create an app that is heavily dependent on videos without being sure that a video has loaded or not.

I'm building an app that is heavily dependent on videos and the only way I'm able to load them is within the `<Video>` component.  Since they never load for me, I'm confused how I will be able to use this npm at all.  I'm looking for alternatives but haven't found one yet.

I'm on SDK48. `"expo-av": "~13.2.1",`

**Comment by RRaideRR** [comment]
> Could this be because expo-av is still running an outdated exoplayer version from 2022? maybe updating the [exoplayer dependency](https://github.com/expo/expo/blob/c9b7fe0ab6600460a94582b2f89824832348c8eb/packages/expo-av/android/build.gradle#L166C34-L166C34) could solve the issue.

Hmm - I don't think that this will solve the root issue because the initial problem occurred on iOS (not quite sure if its also occurring on Android though).

**Comment by hichemfantar** [comment]
> > Could this be because expo-av is still running an outdated exoplayer version from 2022? maybe updating the [exoplayer dependency](https://github.com/expo/expo/blob/c9b7fe0ab6600460a94582b2f89824832348c8eb/packages/expo-av/android/build.gradle#L166C34-L166C34) could solve the issue.
> 
> Hmm - I don't think that this will solve the root issue because the initial problem occurred on iOS (not quite sure if its also occurring on Android though).

Yes that's true but it might fix other playback issues I've been having with certain files on Android.
Also it seems exoplayer has been [deprecated](https://github.com/google/ExoPlayer#deprecation) but updating to the [latest version](https://github.com/google/ExoPlayer/releases) before the deprecation would solve some issues on Android.

**Comment by adriaanbalt** [comment]
I'm also experiencing this issue on SDK48.  I was working on upgrading my app from an old SDK to 48 and now I'm unable to complete the rebuild :(  My previous build worked with expo-av 9.2.3 but something has happened since that version.

**Comment by hichemfantar** [comment]
@adriaanbalt this issue isn't about failing builds, it seems like you have a different issue.
Check the build logs and you should find the cause.
Make sure expo doctor is passing all tests.

**Comment by adriaanbalt** [comment]
My build compiles but the videos do not play.On Oct 30, 2023, at 7:18 AM, Hichem Fantar ***@***.***> wrote:﻿
@adriaanbalt this issue isn't about failing builds, it seems like you have a different issue.
Check the build logs and you should find the cause.
Make sure expo doctor is passing all tests.

**Comment by hichemfantar** [comment]
@adriaanbalt try running `npx expo-doctor@latest` and see if you have incompatible deps.

**Comment by adriaanbalt** [comment]
@hichemfantar 
I've been attempting to upgrade to sdk 49 last night because I noticed it changed `"expo-av": "~13.4.1",`.  But now my app doesn't build at all :(
This is what happens when I run `npx expo-dcotor@latest` even though I've tried to fix some of those CLI issues.

```
npx expo -v
0.10.14
node -v
v16.20.2
npm -v
8.19.4
```

Doctor results:
```
✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check npm/ yarn versions
✖ Check that native modules do not use incompatible support packages
Unexpected error while running 'Check that native modules do not use incompatible support packages' check:
Error: Dependency tree validation for @unimodules/react-native-adapter failed. This validation is only available on Node 16+ / npm 8.
✖ Check for legacy global CLI installed locally
Unexpected error while running 'Check for legacy global CLI installed locally' check:
Error: Dependency tree validation for expo-cli failed. This validation is only available on Node 16+ / npm 8.
✔ Check Expo config (app.json/ app.config.js) schema
✖ Check that native modules use compatible support package versions for installed Expo SDK
Unexpected error while running 'Check that native modules use compatible support package versions for installed Expo SDK' check:
Error: Dependency tree validation for expo-modules-autolinking failed. This validation is only available on Node 16+ / npm 8.
✔ Check that packages match versions required by installed Expo SDK
One or more checks failed, indicating possible issues with the project.
```

**Comment by RRaideRR** [comment]
Hi @adriaanbalt . As @hichemfantar pointed out, your problem has probably nothing to do with this issue. Please open a new issue with a minimal reproducible example and I'm sure the expo team will have a look at it.

**Comment by hichemfantar** [comment]
@adriaanbalt like @RRaideRR said, it's best to open up a new issue so we don't clutter up this one. Send the new link here and I'll try my best to help you in the new issue discussion.

**Comment by adriaanbalt** [comment]
@RRaideRR My problem originated with the fact that my build compiles but the videos never play.  I've been attempting to fix this by decrementing my expo-av version but this doesn't work because it is not compatible with sdk48.  After that I've commenced upgrading to sdk49 because I noticed that the expo-av version is also updated.

The root of my problem is that the videos never load within the <Video> tag and my attempts to fix it are not successful.  

Note that my videos used to load perfectly, the app was good but it was on an old SDK 43 and I want to add new features so I'm forced to upgrade to SDK48 (or 49).

So now I've reverted back to sdk 48, checked that my videos are downloaded to the device, I see the following console logs

```js
 LOG  **onLoad** {"didJustFinish": false, "durationMillis": 10032, "hasJustBeenInterrupted": false, "isBuffering": false, "isLoaded": true, "isLooping": false, "isMuted": false, "isPlaying": false, "pitchCorrectionQuality": "Varispeed", "playableDurationMillis": 10032, "positionMillis": 0, "progressUpdateIntervalMillis": 500, "rate": 1, "shouldCorrectPitch": false, "shouldPlay": false, "target": 1529, "uri": "file:///Users/adriaanbalt/Library/Developer/CoreSimulator/Devices/1B33AEAC-A963-46EF-AFA8-480715DB9674/data/Containers/Data/Application/C06D2779-AD6F-4A34-89BF-9CEEDE628264/Documents/ExponentExperienceData/%2540adriaanbalt%252Fgreatflix/cached-https---storage.googleapis.com-moviemood-meta-all_movies_data_v2-clips-0000001.mp4", "volume": 1}

 LOG  **onPlaybackStatusUpdate** {"didJustFinish": false, "durationMillis": 10032, "hasJustBeenInterrupted": false, "isBuffering": false, "isLoaded": true, "isLooping": false, "isMuted": false, "isPlaying": false, "pitchCorrectionQuality": "Varispeed", "playableDurationMillis": 10032, "positionMillis": 0, "progressUpdateIntervalMillis": 500, "rate": 1, "shouldCorrectPitch": false, "shouldPlay": false, "target": 1529, "uri": "file:///Users/adriaanbalt/Library/Developer/CoreSimulator/Devices/1B33AEAC-A963-46EF-AFA8-480715DB9674/data/Containers/Data/Application/C06D2779-AD6F-4A34-89BF-9CEEDE628264/Documents/ExponentExperienceData/%2540adriaanbalt%252Fgreatflix/cached-https---storage.googleapis.com-moviemood-meta-all_movies_data_v2-clips-0000001.mp4", "volume": 1}
```

But the videos never start to play.

I'm going to continue debugging and will circle back if a solution presents itself.

**Comment by Yormungand** [comment]
@miso-belica Your version was very understandable for me. And it was continuously loading uri till it's working.

And I had to separate props between 2 platforms cuz loadAsync was not working on IOS for me. 

```js

const handleVideoQualityChange = async (item) => {
        setCurrentQuality(item);
        setSheetOpen(false);
        if (Platform.OS === "android") {
            const position = await currentTime;
            await videoRef.current.pauseAsync();
            await videoRef.current?.unloadAsync();
            await videoRef.current?.loadAsync({uri: item?.uri}, {positionMillis: position});
        }
    }

const commonVideoProps = {
        style: [styles.video, { width: window.width, height: window.height }],
        resizeMode: "contain",
        useNativeControls: false,
        shouldPlay: true,
        onPlaybackStatusUpdate: handleVideoStatusUpdate,
        onLoad: handleOnLoad,
        onReadyForDisplay: handleVideoReady,
        pointerEvents: 'none',
    };

    const iosVideoProps = {
        source: {
            uri: currentQuality?.uri,
        },
    };

<PatchedVideo
                        ref={videoRef}
                        {...props}
                        {...commonVideoProps}
                        {...(Platform.OS === 'ios' && iosVideoProps)}
                    />
```

**Comment by libcthorne** [comment]
I'm seeing the same issue on Android. Frequency seems to be random and onError isn't called.

**Comment by alanjhughes** [comment]
`expo-av` is now in maintenance mode. We are working on `expo-video` and `expo-audio` to replace it.

**Comment by hichemfantar** [comment]
@alanjhughes sounds great. Is there an estimation on the release date of the new packages?

**Comment by mihailapuste** [comment]
> `expo-av` is now in maintenance mode. We are working on `expo-video` and `expo-audio` to replace it.

@alanjhughes 
Would there happen to be any betas out for this, so we can test it out. Thanks!

**Comment by ritute** [comment]
We're also experiencing this issue randomly with the error "player error: decoder failed: c2.mtk.avc.decoder" (onError) called.

**Comment by jamiewilbraham** [comment]
Any updates on when we can expect expo-video and expo-audio. TIA

**Comment by brentvatne** [comment]
we're working on those packages for sdk 51, coming in the spring

**Comment by hichemfantar** [comment]
This isn't clear in the documentation but does the current version of expo-av do video caching by default?

**Comment by RRaideRR** [comment]
> This isn't clear in the documentation but does the current version of expo-av do video caching by default?

As far as I know, it does not provide a caching solution. You would need to build something by yourself (e.g. download the assets with expo-filesystem, play them with expo-av, ...)

**Comment by dhairyasenjaliya** [comment]
Hey for those who are facing the `Decoder init failed: OMX.qcom.video.decoder.avc`, it seems like a generic issue when we try to load more than 1 video E.g: when the `<Video>` component is paired with `<Flatlist>` and It pre-renders component as much as you have data (E.g: you have 100 items on Flatlist it pre-renders component to 100 times which makes some video not loaded) the only good solution without caching video is to replace the `<Flatlist>` component to `<BigList>`   which makes sure to render only current index + nextIndex so that `<Video>` did not overwhelm with 100's of data at a time and makes sure once data is scrolled it video loaded without failed for currentIndex + nextIndex   

- Im attaching an Example with `BigList` vs `FlatList` I am still working on the below code but the video is loaded 100% with usage of `BigList` and on`FlatList` some video is not loaded 

**Video Player With BigList**

https://github.com/expo/expo/assets/47522946/a112c80f-b424-4a28-80b9-f514c19ed2d0

**Video Player With FlatList** 

https://github.com/expo/expo/assets/47522946/e99c9c0f-538f-43ab-a3f0-9a2ea413f9c8

**Comment by dhairyasenjaliya** [comment]
Hey for those who are facing the `Decoder init failed: OMX.qcom.video.decoder.avc`, it seems like a generic issue when we try to load more than 1 video E.g: when the `<Video>` component is paired with `<Flatlist>` and It pre-renders component as much as you have data (E.g: you have 100 items on Flatlist it pre-renders component to 100 times which makes some video not loaded) the only good solution without caching video is to replace the `<Flatlist>` component to `<BigList>`   which makes sure to render only current index + nextIndex so that `<Video>` did not overwhelm with 100's of data at a time and makes sure once data is scrolled it video loaded without failed for currentIndex + nextIndex   

- Im attaching an Example with `BigList` vs `FlatList` I am still working on the below code but the video is loaded 100% with usage of `BigList` and on`FlatList` some video is not loaded 

**Video Player With BigList**

https://github.com/expo/expo/assets/47522946/a112c80f-b424-4a28-80b9-f514c19ed2d0

**Video Player With FlatList** 

https://github.com/expo/expo/assets/47522946/e99c9c0f-538f-43ab-a3f0-9a2ea413f9c8

**Note: For Android Try on a Real Device on the Simulator still getting errors**

**Comment by nabww** [comment]
Has anyone figured this out yet, I'm currently facing a challenge playing the videos as well. I get the controls but when I hit play it just blanks out. sdk51.

**Comment by hichemfantar** [comment]
> Has anyone figured this out yet, I'm currently facing a challenge playing the videos as well. I get the controls but when I hit play it just blanks out. sdk51.

Try https://github.com/TheWidlarzGroup/react-native-video and see if that fixes playback issues

**Comment by lit26** [comment]
```jsx
useEffect(() => {
    if (!video?.url) return;
    videoRef.current?.loadAsync({ uri: video.url });
  }, [video?.url]);
```
solve my issue.

**Comment by hirbod** [comment]
Since expo-av is unsupported and we're removing it in SDK54, I am going to close this. Please migrate to expo-video.
